crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,280 @@
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID as u64;
+
+pub const NVTXW3_STREAM_ORDER_INTERLEAVING_NONE: i16 = 0;
+pub const NVTXW3_STREAM_ORDERING_TYPE_UNKNOWN: i16 = 0;
+
+pub fn initialize_simple() -> Result<InterfaceHandle, ResultCode> {
+    let so_name = "libNvtxwBackend.so";
+    let c_mode_string = CString::new(so_name).expect("modeString CString::new failed");
+    let cptr_mode_string = c_mode_string.as_ptr();
+
+    let result;
+    let mut get_interface_func: nvtxw_bindings::nvtxwGetInterface_t = Default::default();
+    unsafe {
+        result = nvtxw_bindings::nvtxwInitialize(
+            NVTXW3_INIT_MODE_LIBRARY_FILENAME,
+            cptr_mode_string,
+            &mut get_interface_func,
+            null_mut(),
+        );
+    }
+
+    if result != NVTXW3_RESULT_SUCCESS {
+        return Err(result);
+    }
+
+    let get_interface = get_interface_func.ok_or(NVTXW3_RESULT_FAILED)?;
+
+    let result;
+    let mut interface_void: *const c_void = null();
+    unsafe {
+        result = get_interface(NVTXW3_INTERFACE_ID_CORE_V1, &mut interface_void);
+    }
+
+    if result != NVTXW3_RESULT_SUCCESS {
+        return Err(result);
+    }
+    if interface_void.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    let ptr_interface = interface_void as *const nvtxw_bindings::nvtxwInterfaceCore_t;
+    let interface;
+    unsafe {
+        interface = *ptr_interface;
+    }
+
+    Ok(interface)
+}
+
+pub fn session_begin_simple(
+    interface: InterfaceHandle,
+    o_output: Option<OsString>,
+    o_merge: Option<OsString>,
+) -> Result<SessionHandle, ResultCode> {
+    let func = interface
+        .SessionBegin
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let c_output: CString;
+    let mut cptr_output: *const c_char = null_mut();
+    if let Some(os_output) = o_output {
+        let s_output = os_output
+            .into_string()
+            .expect("OsString::into_string failed");
+        let session_name = match s_output.strip_suffix(".nsys-rep") {
+            Some(s) => s,
+            None => &s_output,
+        };
+        c_output = CString::new(session_name).expect("CString::new failed");
+        cptr_output = c_output.as_ptr();
+    }
+
+    let c_config: CString;
+    let mut cptr_config: *const c_char = null_mut();
+    if let Some(os_merge) = o_merge {
+        let s_merge = os_merge.to_str().expect("OsString::into_string failed");
+        let s_config = format!("ReportMerge={}", s_merge);
+
+        c_config = CString::new(s_config).expect("CString::new failed");
+        cptr_config = c_config.as_ptr();
```

Comment:
This code style is a bit risky because it's easy to introduce a dangling pointer if the scope of `c_config` later changes. A less risky coding style would be to store the `CString` in a variable as you do now and then when you use it later:

```
configString: c_config.as_ptr()
```

Thus there's no opportunity for that to dangle.

Reply:
Fixed